proposed/0006-unified-structured-data-imporvement.md

```diff
@@ -0,0 +1,147 @@
+
+# Motivation
+
+##Why?
+
+The field `previous_owner_keys` causes code complexity and cofusion. Unified Structured Data can be made to work without it.
+
+##What cases does it support?
+
+This change supports all use of non immutable data (structured data). This covers all non `content only` data on the network and how it is handled.
+
+##Expected outcome
+
+Removing the field `previous_owner_keys` simplifies code and reduces logical complexity during Structured Data Updates and ownership-transfers. This might further simplify vault logic in future to avoid Structured Data Collisions even for the same `location` (64 byte network address for the data) as vaults will now have to look only for a single field (`owners`) and track its changes to make the storage key unique even though `location` wasn't unique.
+
+# Detailed design
+
+##StructuredData
+
+```
+struct StructuredData {
+    tag_type   : u64,
+    identifier : crypto::hash::sha512::Digest,
+    version    : u64,                         // mutable - incrementing (deterministic) version number
+    data       : Vec<u8>,                     // mutable - in many cases this is encrypted
+    owner_keys : vec<crypto::sign::PublicKey> // mutable - n * 32 Bytes (where n is number of owners)
+    signatures : Vec<crypto::sign::Signature> // mutable - detached signature of all the `mutable` fields (barring itself) by owners
+}
+```
+
+Fixed (immutable fields)
+- tag_type
+- identifier
+
+##Flow
+
+The following shows Creation, Updation and Transfer-of-ownership of Structured Data:
+
+**step 0:**
+Client-A sends:
+```
+StructuredData {
+    other_fields: Value-00,
+    owner_keys: A-PublicKey,
+    signatures: Signed by A-PrivateKey,
+}
+```
+ 
+**step 1:**
+Vault receives for the first time. So Just Stores it:
+```
+StructuredData {
+    other_fields: Value-00,
+    owner_keys: A-PublicKey,
+    signatures: Signed by A-PrivateKey,
+}
+```
+ 
+**step 2:**
+Client-A updates it:
+```
+StructuredData {
+    other_fields: Value-11,
+    owner_keys: A-PublicKey,
+    signatures: Signed by A-PrivateKey,
+}
+```
+ 
+**step 3:**
+Vault will check if new data is sent by valid owner. Use `owner_keys` from the stored data in step 1 to verify signature of data in step 2. If valid replace previous data and store new one:
+```
+StructuredData {
+    other_fields: Value-11,
+    owner_keys: A-PublicKey,
+    signatures: Signed by A-PrivateKey,
+}
+```
+ 
+**step 4:**
+Client-A wants to transfer ownership of Structured Data to Client-B:
+```
+StructuredData {
+    other_fields: Value-22,
+    owner_keys: B-PublicKey,
+    signatures: Signed by A-PrivateKey,
+}
+```
+ 
+**step 5:**
+Vault will check if new data is sent by valid owner. Use `owner_keys` from the stored data in step 3 to verify signature of data in step 4. If valid replace previous data and store new one:
+```
+StructuredData {
+    other_fields: Value-22,
+    owner_keys: B-PublicKey,
+    signatures: Signed by A-PrivateKey,
+}
```

Comment:
This may be where we get an issue. at the routing level checks will be made on this data for now a `Put` just needs to check at least one owner has signed the `Put` message (routing message).
  
For updates (`Post`) though in the routing layer then routing will check that the routing message is signed by one of the `PreviousOwners`  at the moment and pass the message up to vaults. If routing did not do this check then this scheme here would work, but routing may pass up invalid data? 
At the moment though it can be argued routing does only a basic check not a full check (it only checks a single signature) and the data may still fail in upper layers, however it does at least check this data is not being spammed by a non owner.